### PR TITLE
[ONNX] Supports scatter_add with different static shape of src and index

### LINKED ISSUE #89190: [ONNX] OnnxExporterError when exporting SAGEConv with dynamic_axes
### 🐛 Describe the bug

I'm exporting a model built with `torch_geometric.nn.SAGEConv`.
As different graph may have different number of nodes and edges, I make the first dimension of `x` and the second dimension of `edge_index` dynamic. Then the error occurs. 

```python
import torch
from torch_geometric.nn import SAGEConv

class MyModel(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = SAGEConv(8, 16)

    def forward(self, x, edge_index):
        x = self.conv1(x, edge_index)
        return x

model = MyModel()
x = torch.randn(3, 8)
edge_index = torch.tensor([[0, 1, 1, 2], [1, 0, 2, 1]])

torch.onnx.export(
    model,
    (x, edge_index),
    "model.onnx",
    opset_version=16,
    input_names=["x", "edge_index"],
    dynamic_axes={
        "x": {0: "n_nodes"},
        "edge_index": {1: "n_edges"}
    }
)
```

```
torch.onnx.errors.OnnxExporterError: Unsupported: ONNX export of operator scatter_add, `index` ([None, None]) should have the same dimensionality as `src` ([None, 8]). Please feel free to request support or submit a pull request on PyTorch GitHub: https://github.com/pytorch/pytorch/issues
```

Without the `dynamic_axes`, the model can be exported sucessfully, however it can't infer on other graphs with different number of nodes or edges.

### Versions

```
torch==1.13.0
torch-geometric==2.1.0.post1
torch-scatter==2.0.9
torch-sparse==0.6.15
```

### PR DESCRIPTION
**Pull request by titaiwangms**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89787

Prior to this change, the converter doesn't support `scatter_add` with different shape of `src` and `index`, while [it's claimed to be supported by PyTorch](https://pytorch.org/docs/stable/generated/torch.Tensor.scatter_add_.html#torch.Tensor.scatter_add_) in a way that scatter shape would be accommodated to index shape. This PR adds `onnx::Slice` to adjust the shape of `src` when a static and mismatched shape is found. However, if both of the shape (src and index) is set to dynamic, they are expected to be the same shape from ONNX due to the spec. More ScatterElements details on https://github.com/onnx/onnx/issues/4672

### CODE REVIEW

File: torch/onnx/symbolic_opset16.py [hunk 1]
```diff
@@ -75,12 +75,47 @@ def scatter_add(g: jit_utils.GraphContext, self, dim, index, src):
...
+            if d is None:
+                # the axe with dynamic shape is ignored, and will be aligned by
+                # setType later
+                continue
+            if src_sizes[idx] < d:
+                return symbolic_helper._unimplemented(
+                    "scatter_add",
+                    f"`index` ({index_sizes}) should have smaller or equal (<=) size at any dimension than `src` ({src_sizes})",
+                )
+            axes.append(idx)
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset16.py:103]: Should this happen only when `src_sizes[idx] > d`? No need to slice when they are equal right?

File: torch/onnx/symbolic_opset16.py [hunk 2]
```diff
@@ -75,12 +75,47 @@ def scatter_add(g: jit_utils.GraphContext, self, dim, index, src):
...
+            ends.append(d)
+        starts = [0] * len(ends)
+        if axes and starts and ends:
+            src = symbolic_helper._slice_helper(
+                g, src, axes=axes, starts=starts, ends=ends
+            )
+        # Align the dynamic sizes of src and index
+        # NOTE: Even if users set src and index with different dynamic axes, they are
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset16.py:111]: It looks like symbolic function does not have enough information to remediate if shape don't match under dynamic size. 
Does setting sizes manually below has any effect? (Is it to pass onnx checker?) If not I'd suggest remove it and comment above why mismatch under dynamic shape cannot be fixed here.

**Review by BowenBao — APPROVED** [review]
LGTM, a few suggestions.

File: torch/onnx/symbolic_opset16.py (hunk 1 above)
Comment (titaiwangms) [reply torch/onnx/symbolic_opset16.py:110]: Done

File: torch/onnx/symbolic_opset16.py (hunk 2 above)
Comment (titaiwangms) [reply torch/onnx/symbolic_opset16.py]: Done

**Pull request closed without merging**